File: readability.c
#include <ctype.h>
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <math.h>

//function to count letters
int count_letters(string text);
//function to count words
int count_words(string text);
//function to count sentences
int count_sentences(string text);
int main(void)
{
    //prompt for text
    string text = get_string("Text: ");
    //average number of letters will be the quotient of letters divided by words times 100
    //here and in all calculations requiring precision, I multiplied by a floating point constant
    //in the numerator and denominator, otherwise the division would yield an int quotient
    //which does not allow for decimals
    float avg_letters = (100.0000 * ((1.00000 * count_letters(text)) / ((1.00000 * count_words(text))))); 
    float avg_sentences = (100.0000 * ((1.00000 * count_sentences(text)) / ((1.00000 * count_words(text)))));
    //applying math of index
    int cl_index = round((.0588 * avg_letters) - (.296 * avg_sentences) - 15.8); 
    if (cl_index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (cl_index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", cl_index);

    }
}

int count_sentences(string text)
{
    //initialize a count for number of sentences at 0
    int sentence_count = 0;
    //iterate this loop for every character in the string
    for (int k = 0; k < strlen(text); k++)
    {
        //move all chars, adding one to sentence count if the value
        //equals the ascii value for . ? !
        if (text[k] == 46 || text[k] == 33 || text[k] == 63)
        {
            sentence_count++;
        }
    }
    //return the number of sentences
    return sentence_count;
}

int count_letters(string text)
{
    //initialize count for number of letters at 0
    int letter_count = 0;
    //repeat loop for the whole sentence
    for (int i = 0; i < strlen(text); i++)
    {
        //if a character is not whitespace and is in alphabet 
        //(so not apostrophes in contractions, for exammple,
        //then add 1 to letter count
        if (isspace(text[i]) == 0 && isalpha(text[i]) != 0)
        {
            letter_count++;
        }
    }
    // return letter count
    return letter_count;
}

int count_words(string text)
{
    //initialize count at 1 (because there's always one more word than spaces
    int word_count = 1;
    //iterate through whole sentence, if isspace yields a nonzero value (indincating
    //white space), then add to the word counter
    for (int j = 0; j < strlen(text); j++)
    {
        if (isspace(text[j]) != 0)
        {
            word_count++;
        }
    }
    //return word_count
    return word_count;
}

File: readability.py
# Sources that helped me out:

# https://stackoverflow.com/questions/54134072/how-to-re-prompt-a-user-after-an-invalid-input

# https://www.geeksforgeeks.org/floor-ceil-function-python/

# https://www.geeksforgeeks.org/how-to-convert-float-to-int-in-python/

# https://stackoverflow.com/questions/1155617/count-the-number-of-occurrences-of-a-character-in-a-string

# https://stackoverflow.com/questions/24878174/how-to-count-digits-letters-spaces-for-a-string-in-python

# https://www.geeksforgeeks.org/formatted-string-literals-f-strings-python/


from cs50 import get_string


def main():

    text = get_string("Text: ")

    # calculate average letters and sentences using the functions definfed below

    avg_letters = (100.0000 * ((1.00000 * count_letters(text)) / ((1.00000 * count_words(text)))))
    avg_sentences = (100.0000 * ((1.00000 * count_sentences(text)) / ((1.00000 * count_words(text)))))
    cl_index = round((.0588 * avg_letters) - (.296 * avg_sentences) - 15.8)

    # make necessary conditionals to print
    if cl_index < 1:
        print("Before Grade 1")
    elif cl_index >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {cl_index}")


def count_sentences(x):

    #learned this function from 4th link above

    s_count = x.count('.') + x.count('!') + x.count('?')
    return s_count


def count_letters(x):

    # using isalpha function to count number of letters per sentence, adding 1 to a counter for every character that's a letter
    l_count = 0
    for i in range(len(x)):
        if x[i].isalpha():
            l_count = l_count + 1
    return l_count


def count_words(x):

    # same as above, just with isspace(). Using counter starting at 1 because an input like "Hello." needs to register as a sentence
    # despite having no spaces
    w_count = 1
    for i in range(len(x)):
        if x[i].isspace():
            w_count = w_count + 1
    return w_count
    

main()